lib/events/azsessions/azsessions.go

```diff
@@ -0,0 +1,549 @@
+		c.ServiceURL.Scheme = "http"
+	}
+
+	params, err := url.ParseQuery(c.ServiceURL.EscapedFragment())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	c.ServiceURL.Fragment = ""
+	c.ServiceURL.RawFragment = ""
+
+	c.ClientID = params.Get(clientIDFragParam)
+
+	return nil
+}
+
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Log == nil {
+		c.Log = logrus.WithField(trace.Component, teleport.SchemeAZBlob)
+	}
+
+	return nil
+}
+
+func NewHandler(ctx context.Context, cfg Config) (*Handler, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var cred azcore.TokenCredential
+	if cfg.ClientID != "" {
+		c, err := azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
+			ID: azidentity.ClientID(cfg.ClientID),
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		cred = c
+	} else {
+		c, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		cred = c
+	}
+
+	cred = &cachedTokenCredential{TokenCredential: cred}
+
+	service, err := azblob.NewServiceClient(cfg.ServiceURL.String(), cred, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	session, err := service.NewContainerClient(sessionContainerName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	inprogress, err := service.NewContainerClient(inprogressContainerName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if _, err := cErr2(session.GetProperties(ctx, nil)); err != nil {
+		if !trace.IsNotFound(err) && !trace.IsAccessDenied(err) {
+			return nil, trace.Wrap(err)
+		}
+		cfg.Log.WithError(err).Debug("Failed to confirm that the " + sessionContainerName + " container exists, attempting creation.")
+		// someone else might've created the container between GetProperties and
+		// Create, so we ignore AlreadyExists
+		if _, err := cErr2(session.Create(ctx, nil)); err != nil && !trace.IsAlreadyExists(err) {
+			if !trace.IsAccessDenied(err) {
+				return nil, trace.Wrap(err)
+			}
+			cfg.Log.WithError(err).Warn(
+				"Could not create the " + sessionContainerName + " container, please ensure it exists or session recordings will not be stored correctly.")
+		}
+	}
+
+	if _, err := cErr2(inprogress.GetProperties(ctx, nil)); err != nil {
+		if !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err)
+		}
+		cfg.Log.WithError(err).Debug("Failed to confirm that the " + inprogressContainerName + " container exists, attempting creation.")
+		// someone else might've created the container between GetProperties and
+		// Create, so we ignore AlreadyExists
+		if _, err := cErr2(inprogress.Create(ctx, nil)); err != nil && !trace.IsAlreadyExists(err) {
+			if !trace.IsAccessDenied(err) {
+				return nil, trace.Wrap(err)
+			}
+			cfg.Log.WithError(err).Warn(
+				"Could not create the " + inprogressContainerName + " container, please ensure it exists or session recordings will not be stored correctly.")
+		}
+	}
```

Comment:
More of a nit, but you could probably factor this out into a method like `ensureContainer` and reuse it for both "session" and "inprogress" containers provisioning to avoid some duplication.

Reply:
Will do.

Reply:
6fb5c2183 (after rebase)